File: python/is_subsequence.py
class Solution:
    # time complexity: O(n)
    def isSubsequence(self, s: str, t: str) -> bool:
        i: int = 0
        j: int = 0

        while i < len(s) and j < len(t):
            if s[i] == t[j]:
                i += 1
            j += 1

        return i == len(s)


def test_isSubsequence_case_1():
    # arrange
    s: str = "abc"
    t: str = "ahbgdc"
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.isSubsequence(s, t)

    # assert
    assert expected == actual

File: rust/src/code/is_subsequence.rs
struct Solution {}

impl Solution {
    pub fn is_subsequence(s: String, t: String) -> bool {
        let s: Vec<char> = s.chars().collect();
        let t: Vec<char> = t.chars().collect();
        let mut i: usize = 0;
        let mut j: usize = 0;
        while i < s.len() && j < t.len() {
            if s[i] == t[j] {
                i += 1;
            }
            j += 1;
        }

        i == s.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn is_subsequence_case_1() {
        // arrange
        let s: String = "abc".to_string();
        let t: String = "ahbgdc".to_string();
        let expected: bool = true;

        // act
        let actual = Solution::is_subsequence(s, t);

        // assert
        assert_eq!(expected, actual);
    }
}
